[1단계 - 페이먼츠 모듈] 기린(정유정) 미션 제출합니다.

jeongyou opened the pull request:
안녕하세요 마르코~ 이번에 리뷰를 받게된 기린이라고 합니다😊 
만나서 반갑습니다~ 
리뷰 잘 부탁드립니다!

## 🎯 페이먼츠 모듈

이번 미션을 통해 다음과 같은 학습 경험들을 쌓는 것을 목표로 합니다.

### 1단계 - Module

- 재사용 가능한 모듈화된 컴포넌트를 개발하고 npm에 배포할 수 있다.
- 재사용 가능한 커스텀 훅을 개발하고 npm에 배포할 수 있다.
- Storybook을 사용하여 컴포넌트의 다양한 상태를 시각적으로 문서화하고 테스트할 수 있다.
- React Testing Library를 사용하여 컴포넌트와 커스텀 훅에 대해 폭넓은 테스트를 작성할 수 있다.

## 🕵️ 셀프 리뷰(Self-Review)

### 제출 전 체크 리스트

- [x] 기능 요구 사항(모달 컴포넌트 및 커스텀 훅 개발, npm 배포)을 모두 구현했고, 정상적으로 동작하는지 확인했나요?
- [x] Storybook과 RTL 테스트 케이스를 모두 작성했나요?
- [x] 코드 컨벤션을 지키고, 에러 핸들링을 고려했나요?
- [x] 배포한 npm 패키지가 정상적으로 설치 및 작동하나요?
  - 배포 npm 링크: 
  - 컴포넌트: https://www.npmjs.com/package/kirin-modal-component
  - 훅: https://www.npmjs.com/package/kirin-payment-hook
- [x] Storybook 데모 페이지에 정상적으로 접근할 수 있나요?
  - Storybook 링크: https://68130cd34b90e54b5588aff6-nhkldiehce.chromatic.com/?path=/docs/modal--docs

### 리뷰 요청 & 논의하고 싶은 내용

### 1) 이번 단계에서 가장 많이 고민했던 문제와 해결 과정에서 배운 점


1. **모달 재사용성을 위한 합성 컴포넌트**
모달 컴포넌트의 재사용성과 구조적인 확장성을 어떻게 설계할지에 대해 가장 많이 고민했습니다.
처음에는 `header`, `title`, `body`, `footer`에 들어가는 내용을 모두 `props`로 받아서 처리하려 했지만, 이렇게 구성하면 모달 내부 구조가 복잡해지고 재사용하기 어려운 구조가 된다는 문제가 있었습니다. <br/>
그래서 이를 해결하기 위해 합성 컴포넌트(Compound Component) 패턴을 적용했습니다. `<Modal.Header>`, `<Modal.Body>` 같은 방식으로 모달 내부 구성요소를 자식으로 선언적으로 조립할 수 있게 하고, `Modal` 컴포넌트는 최소한의 `props`만 받도록 설계했습니다.


2. **`Context API`를 활용한 `onHide` 상태 관리 개선**
이번 단계에서 가장 많이 고민했던 부분은 모달을 닫는 함수인 `onHide`를 어떻게 효율적으로 전달할 것인가였습니다. 처음에는 단순히 `props`를 백그라운드가 있는 상위 모달 컴포넌트와 X 버튼이 있는 하위 컴포넌트에 각각에 `onHide`를 넘겨주는 방식으로 구현했습니다. 하지만 이러한 방식은 컴포넌트가 복잡해질수록 `props` 전달이 중첩되고 가독성이 떨어진다는 문제점이 있었습니다.<br/>
그래서 저희는 "상위 모달에서 받은 `onHide`를 하위 컴포넌트로 자연스럽게 공유할 수는 없을까?"라는 의문을 가졌고, 이를 해결하기 위해 `Context API`를 도입했습니다. 특히 합성 컴포넌트 패턴과 함께 사용하면, 상위에서 제공한 `context` 값을 하위 요소들이 별도의 `props` 전달 없이도 쉽게 사용할 수 있다는 점에서 큰 장점을 느꼈습니다.<br/>
결과적으로 `onHide`를 `Context`로 관리하고, 하위의 `Modal.Title` 컴포넌트 내부의 X 버튼에서도 이를 사용할 수 있도록 구현함으로써, 컴포넌트를 사용하는 입장에서는 더 적은 `props`로 간결하게 사용할 수 있었고, 코드의 가독성과 유지보수성도 높아졌습니다. 이 과정을 통해 합성 컴포넌트와 `Context API`를 결합한 구조의 장점을 실감할 수 있었고, 보다 유연하고 확장성 있는 컴포넌트 설계에 대한 이해를 깊이 있게 할 수 있었습니다.


3. **SVG 이미지 배포 이슈**
처음에는 닫기 버튼 아이콘을 `lib/assets` 폴더에 `.svg` 파일로 저장한 뒤 `img` 태그로 불러와 사용했습니다.
하지만 해당 컴포넌트를 NPM 패키지로 배포한 뒤 다른 프로젝트에서 설치해보니, 이미지 파일이 번들링되지 않아 닫기 아이콘이 보이지 않는 문제가 발생했습니다. 이미지가 함께 배포되도록 설정을 변경해보기도 했지만, 환경에 따라 제대로 작동하지 않아 안정적인 해결이 어려웠습니다. 그래서 고민 끝에 배포 시 파일 누락을 방지하기 위해, `SVG` 태그를 코드에 직접 삽입하는 방식으로 변경했습니다.<br/>
> :question:PR을 작성하던 중 한 크루가 이미지 배포 문제를 해결하는 방법을 공유해 주었는데, 저희는 현재 SVG 태그를 직접 삽입하는 방식도 괜찮은 해결책이라고 생각하고 있습니다. 혹시 마르코는 이 방식에 대해 어떻게 생각하는지 궁금합니다...! 만약 더 바람직한 방향이 있다면 리뷰 과정에서 반영해보려고 합니다.


4. **`useValidation` 훅 설계 과정**
훅을 설계하기에 앞서, 이전 단계였던 페이먼츠 미션에서 사용한 코드를 서로 공유하고 함께 읽어보는 시간을 가졌습니다. 그 과정에서 각자 구현한 유효성 검사 훅의 내부 로직이 상당히 유사하다는 것을 알게 되었습니다. 반환값도 모두`error`상태, `errorMessage`, 전체 에러 여부, 유효성 검사 함수 등으로 일관되게 구성되어 있었습니다. 이 공통점을 바탕으로 훅 설계와 모듈화 작업은 비교적 수월하게 진행할 수 있었고, 모달 기능에 비해 빠르게 구현을 마칠 수 있었습니다.


---

### 2) 이번 리뷰를 통해 논의하고 싶은 부분

1. **모달 컴포넌트 자유도**
모달 컴포넌트를 제작할 때, 사용자에게 어느 정도의 자유도를 제공해야 할지에 대해 고민이 많았습니다. 처음에는 레이아웃과 스타일을 고정하고, 버튼이나 색상 등도 내부에서 모두 제어하려 했습니다. 하지만 그렇게 되면 사용자 입장에서 확장성이 떨어지고, 오히려 직접 모달을 만드는 편이 더 낫다고 느낄 수도 있겠다는 생각이 들었습니다.<br/>
반대로 너무 많은 자유도를 주면 `props`의 수가 급격히 늘어나고, 코드가 복잡해질 것 같았습니다. 그래서 저희는 필수 `props`만 받되, 나머지 영역은 `children`과 합성 컴포넌트 방식으로 구성하여 사용자에게 구조적 유연성을 제공하는 방식으로 설계했습니다.
모달 컴포넌트를 설계할 때, 어느 정도의 자유도를 허용하는 것이 바람직한지 궁금합니다. 


2. **CVC와 비밀번호 유효성 훅**
처음 훅을 설계할 때, 총 4개의 입력 필드에 대해 각각의 유효성 검사 훅을 만들기로 했습니다.<br/>
- useCardNumberValidation (카드 번호)
- useExpirationDateValidation (유효기간)
- useCvcNumberValidation (CVC)
- usePasswordValidation (비밀번호)<br/>
각 훅은 해당 필드의 에러 상태를 관리하는 역할을 했습니다. 예를 들어, 카드 번호는 4개의 입력칸이 필요해 boolean 4개로 에러 상태를 관리했고, 유효기간은 월/년 2개로 나누어 관리했습니다.<br/>
하지만 CVC와 비밀번호는 단일 입력값이기 때문에 각각의 훅 내부에서 동일한 로직과 동일한 방식으로 boolean 하나만으로 에러 상태를 관리하고 있었습니다. 결국 두 훅은 이름만 다르고 내부 로직은 100% 동일한 구조가 되어버렸습니다.<br/>
이런 경우, 훅을 구분해서 각각 따로 유지하는 것이 명확한 책임 분리를 위한 좋은 선택일지, 혹은 단일 훅으로 통일하여 재사용성을 높이는 것이 더 나은 선택일지 궁금합니다. 마르코는 어떻게 생각하실까요? 

---

## ✅ 리뷰어 체크 포인트

<!-- 리뷰어가 이 PR을 검토할 때 중점적으로 확인할 사항입니다.
코드의 완성도뿐만 아니라, 리뷰이가 구현 과정에서 어떤 고민과 결정을 하며 학습했는지도 함께 고려해 주세요. -->

### 1. 모달 컴포넌트의 유연성

- 다양한 위치(중앙, 하단 등)와 내용을 prop으로 받아 모달을 유연하게 구성할 수 있는가?
- 열기, 닫기, 확인 등의 사용자 정의 이벤트 핸들러가 자연스럽게 동작하는가?
- Storybook을 통해 다양한 시나리오를 명확히 테스트했는가?

### 2. 커스텀 훅의 역할과 모듈화

- 카드 정보 입력 유효성 검사 훅이 각각 명확한 책임을 가지도록 분리되어 있는가? 
- 훅이 유효성 검사 결과와 에러 메시지를 명확히 반환하는가?
- 독립적인 테스트 케이스를 작성하여 정상/비정상 입력 모두를 폭넓게 검증했는가?

### 3. 모듈 배포
- npm 배포를 위한 설정이 적절히 구성되어 있는가?

jeongyou committed:
docs: 기능 명세서 작성

Co-authored-by: 호이초이 <hoyyChoi@users.noreply.github.com>

jeongyou committed:
feat: Modal component 제작

jeongyou committed:
test: Modal component 스토리북 제작

jeongyou committed:
chore: emotion 설치

jeongyou committed:
chore: 페어 서버 연결

jeongyou committed:
feat: Modal.Header 제작

jeongyou committed:
feat: Modal.Body, Footer, Title 제작

jeongyou committed:
style: prettier 설정 추가

jeongyou committed:
docs: 기능 명세서 수정

jeongyou committed:
style: Modal component background 스타일 적용

jeongyou committed:
style: 브라우저 기본 css reset

jeongyou committed:
chore: emotion react 적용 설정

jeongyou committed:
refactor: background 타입 옵셔널 지정

jeongyou committed:
test: 스토리북 background props 추가

jeongyou committed:
feat: Modal Container position 동적 할당

Co-authored-by: 호이초이 <hoyyChoi@users.noreply.github.com>

jeongyou committed:
feat: Modal.header의 close 버튼과 children 스타일 추가

jeongyou committed:
feat: Modal의 영역을 전체 화면으로 수정

jeongyou committed:
feat: Modal.Title의 font, color 스타일 (color 동적)

jeongyou committed:
feat: Modal component에 show, onHide props 추가

Co-authored-by: 호이초이 <hoyyChoi@users.noreply.github.com>

jeongyou committed:
feat: Modal Container에 gap 속성 추가

Co-authored-by: 호이초이 <hoyyChoi@users.noreply.github.com>

jeongyou committed:
feat: Modal Component 타입스크립트 적용

Co-authored-by: 호이초이 <hoyyChoi@users.noreply.github.com>

jeongyou committed:
test: Modal Component 스토리북 제작

Co-authored-by: 호이초이 <hoyyChoi@users.noreply.github.com>

jeongyou committed:
chore: Modal Component type 주석 추가

Co-authored-by: 호이초이 <hoyyChoi@users.noreply.github.com>

jeongyou committed:
chore: 모달 컴포넌트 type export

Co-authored-by: 호이초이 <hoyyChoi@users.noreply.github.com>

jeongyou committed:
build: d.ts 생성 설정

Co-authored-by: 호이초이 <hoyyChoi@users.noreply.github.com>

jeongyou committed:
feat: 모달 닫기 기능 추가. ESC + Backdrop

Co-authored-by: 호이초이 <hoyyChoi@users.noreply.github.com>

jeongyou committed:
docs: 배포용 readme 작성

jeongyou committed:
docs: 기능 명세서 작성

jeongyou committed:
docs: 기능 명세서 수정

Co-authored-by: 호이초이 <hoyyChoi@users.noreply.github.com>

jeongyou committed:
test: useCardNumberValidation 훅 테스트 코드 작성

Co-authored-by: 호이초이 <hoyyChoi@users.noreply.github.com>

jeongyou committed:
feat: useCardNumberValidation 훅 구현

Co-authored-by: 호이초이 <hoyyChoi@users.noreply.github.com>

jeongyou committed:
refactor: 공용함수 분리

jeongyou committed:
test: useExpirationDateValidation 훅 테스트 코드 작성

Co-authored-by: 호이초이 <hoyyChoi@users.noreply.github.com>

jeongyou committed:
feat: useExpirationDateValidation 훅 구현

Co-authored-by: 호이초이 <hoyyChoi@users.noreply.github.com>

jeongyou committed:
feat: 월, 년 검증 함수 구현

Co-authored-by: 호이초이 <hoyyChoi@users.noreply.github.com>

jeongyou committed:
test: useCvcNumberValidation 테스트 코드 작성

Co-authored-by: 호이초이 <hoyyChoi@users.noreply.github.com>

jeongyou committed:
feat: useCvcNumberValidation 훅 구현

Co-authored-by: 호이초이 <hoyyChoi@users.noreply.github.com>

jeongyou committed:
test: usePasswordValidation 테스트 코드 작성

Co-authored-by: 호이초이 <hoyyChoi@users.noreply.github.com>

jeongyou committed:
feat: usePasswordValidation 훅 구현

Co-authored-by: 호이초이 <hoyyChoi@users.noreply.github.com>

jeongyou committed:
feat: 훅 및 함수 매개변수 및 return 타입 추가

Co-authored-by: 호이초이 <hoyyChoi@users.noreply.github.com>

jeongyou committed:
chore: 폴더 구조 변경

Co-authored-by: 호이초이 <hoyyChoi@users.noreply.github.com>

jeongyou committed:
refactor: validation isExpirationDate 함수 리펙터링

Co-authored-by: 호이초이 <hoyyChoi@users.noreply.github.com>

jeongyou committed:
feat: constants 처리

Co-authored-by: 호이초이 <hoyyChoi@users.noreply.github.com>

jeongyou committed:
refactor: 일반 변수를 상수로 처리

Co-authored-by: 호이초이 <hoyyChoi@users.noreply.github.com>

jeongyou committed:
refactor: 훅 및 함수에 타입 적용

Co-authored-by: 호이초이 <hoyyChoi@users.noreply.github.com>

jeongyou committed:
chore: 유효성 검사 훅 export를 위한 index 파일 추가

Co-authored-by: 호이초이 <hoyyChoi@users.noreply.github.com>

jeongyou committed:
docs: npm 명세서 추가

Co-authored-by: 호이초이 <hoyyChoi@users.noreply.github.com>

jeongyou committed:
docs: 기능 명세서 수정

Co-authored-by: 호이초이 <hoyyChoi@users.noreply.github.com>

jeongyou committed:
test: 모달 컴포넌트 스토리 수정

- onHide 작동 되도록

Co-authored-by: 호이초이 <hoyyChoi@users.noreply.github.com>

jeongyou committed:
build: 스토리북 배포

- onHide 작동 되도록

Co-authored-by: 호이초이 <hoyyChoi@users.noreply.github.com>

jeongyou committed:
build: 스토리북 재배포

Co-authored-by: 호이초이 <hoyyChoi@users.noreply.github.com>

jeongyou committed:
build: 스토리북 배포

Co-authored-by: 호이초이 <hoyyChoi@users.noreply.github.com>

jeongyou committed:
build: 스토리북 재배포

Co-authored-by: 호이초이 <hoyyChoi@users.noreply.github.com>

jeongyou committed:
build: 스토리북 배포

Co-authored-by: 호이초이 <hoyyChoi@users.noreply.github.com>

imakerjun reviewed (commented):
@coderabbitai Based on commit history, identify the **recurring work patterns** this developer uses to solve problems.

Focus on these two key questions:

1. **What consistent workflow or cycle does the developer follow?**
    
    – For example: feature → styling → testing. Explain with specific commit examples.
    
2. **What does this pattern suggest about the developer’s mindset or strategy?**
    
    – e.g., feature-first, cleanup-later, test-driven, etc.
    

Structure the result with a **clear title + example + interpretation**, and provide the answer **in Korean**.

wmakerjun commented:
@coderabbitai Based on commit history, identify the **recurring work patterns** this developer uses to solve problems.

Focus on these two key questions:

1. **What consistent workflow or cycle does the developer follow?**
    
    – For example: feature → styling → testing. Explain with specific commit examples.
    
2. **What does this pattern suggest about the developer’s mindset or strategy?**
    
    – e.g., feature-first, cleanup-later, test-driven, etc.
    

Structure the result with a **clear title + example + interpretation**, and provide the answer **in Korean**.

wonsss commented on components/src/lib/Modal.tsx:
```diff
@@ -0,0 +1,181 @@
+  );
+};
+
+Modal.Header = ({ closeButton = false, children }: ModalHeaderProps) => {
```
모달 컴포넌트가 재사용 가능한 형태로 잘 구현되어 있네요.
조립할 수 있도록 컴파운드 컴포넌트 패턴을 잘 적용해주셨네요 👍 
해당 패턴을 사용함으로써 어떤 점이 좋다고 생각하시나요?

jeongyou replied:
이번에 처음으로 컴파운드 컴포넌트 패턴이라는 개념을 알게 되었고, 직접 적용까지 해보면서 많은 걸 느꼈습니다!

저는 부모 컴포넌트에서 여러 props를 일일이 넘겨줘서 모달 내부 구조를 제어하는 방식만 알고 있었는데, 이 방식은 컴포넌트가 커질수록 props가 복잡해지고 각 영역(헤더, 바디, 푸터)의 역할이 섞이기 쉬웠던 것 같습니다.

반면에 컴파운드 컴포넌트 패턴을 적용해보니까 Modal.Header, Modal.Body, Modal.Footer처럼 기능 단위로 역할을 명확히 분리하면서 선언적으로 조립할 수 있는 구조가 만들어졌습니다. 이 방식은 사용자의 입장에서도 UI 구조를 더 쉽게 파악할 수 있고, 유지보수나 재사용성 측면에서도 확실히 더 유리하다고 느꼈습니다.

그리고 저희는 context를 활용해 닫기 함수 같은 공통 동작을 하위 컴포넌트로 공유할 수 있게 만들었는데, 불필요한 props 전달도 줄일 수 있었던 점이 좋았습니다!

wonsss replied:
잘 이해하셨네요! 멋집니다 👍

wonsss commented on components/src/lib/Modal.tsx:
```diff
@@ -0,0 +1,181 @@
+
+const ModalFooterStyle = css`
+  width: 100%;
+`;
+
+const ModalCloseStyle = css`
+  cursor: pointer;
+`;
+
+const getPositionStyle = (position: string) => {
+  switch (position) {
+    case "center":
+      return css`
+        width: calc(100% - 72px);
+        border-radius: 8px;
+        position: absolute;
+        top: 50%;
+        left: 50%;
+        transform: translate(-50%, -50%);
+      `;
+    case "bottom":
+      return css`
+        width: 100%;
+        border-radius: 8px 8px 0 0;
+        position: fixed;
+        bottom: 0;
+        left: 0;
+      `;
+    case "top":
+      return css`
+        width: 100%;
+        border-radius: 0 0 8px 8px;
+        position: fixed;
+        top: 0;
+        left: 0;
+      `;
+    default:
+      return css``;
+  }
+};
```
스타일 관련 코드가 컴포넌트 로직과 같은 파일에 있어 가독성이 떨어질 수 있어요.
스타일 관련 코드를 별도의 파일(ex, Modal.styles.ts)로 분리해보면 어떨까요?

jeongyou replied:
코드를 작성할 때는 몰랐는데 스타일 코드가 엄청나게 길어졌네요...! 이 부분도 가독성과 관심사 분리를 위해 다른 파일로 분리하였습니다!

관련 커밋:

wonsss commented on components/src/lib/useKeyEscClose.ts:
```diff
@@ -0,0 +1,15 @@
+import { useEffect } from "react";
+
+const useKeyEscClose = (onHide: () => void) => {
+  useEffect(() => {
+    const escKeyModalClose = (e: KeyboardEvent) => {
+      if (e.key === "Escape") {
+        onHide();
+      }
+    };
+    window.addEventListener("keydown", escKeyModalClose);
+    return () => window.removeEventListener("keydown", escKeyModalClose);
+  }, []);
+};
+
+export default useKeyEscClose;
```
커스텀 훅을 사용하여 키보드 이벤트 로직을 잘 분리하셨네요! 👍

wonsss commented on components/src/lib/useKeyEscClose.ts:
```diff
@@ -0,0 +1,15 @@
+    };
+    window.addEventListener("keydown", escKeyModalClose);
+    return () => window.removeEventListener("keydown", escKeyModalClose);
+  }, []);
```
의존성 배열을 비우신 의도가 있으실까요?

jeongyou replied:
이 부분은 사실 별 생각 없이 의존성 배열을 비워두었는데요, 페어와 이야기해보면서 useKeyEscClose 훅의 useEffect 의존성 배열이 비어 있으면 최초 렌더링 시점의 onHide만 참조하게 되는 문제를 발견했습니다.

만약 onHide가 리렌더링될 때 새로 정의되는 함수이거나 내부에서 상태 값을 참조하고 있다면, 그 변화가 반영되지 않아 예상치 못한 동작이 생길 수 있다고 생각해서, useEffect의 의존성 배열에 onHide를 포함하도록 수정해보았습니다!

관련 커밋:

❓ 그리고 한가지 궁금한 점이 있습니다!
지금은 문제없어 보여도, 잠재적인 위험을 예방하기 위해 의존성을 명시적으로 넣어주는 습관을 들이는 것이 좋은지 궁금합니다! 마르코는 이런 경우, 현 시점에 문제가 없더라도 가능성을 고려해서 의존성을 포함시키는 편인지 궁금합니다!

wonsss replied:
useEffect 훅에서 의존성 배열을 명시적으로 지정하는 것은 매우 중요합니다. 이는 사이드 이펙트의 실행 시점을 정확하게 제어하고, 예기치 않은 동작이나 성능 문제를 방지하기 위함입니다.
reactive 의존성을 명시하는 것도 중요하며, 그 외에 불필요한 의존성은 제거하는 것도 중요합니다.
자세한 내용은 아래 공식 문서에서 확인해보시면 도움이 되실거예요.

https://react.dev/reference/react/useEffect#specifying-reactive-dependencies
https://react.dev/reference/react/useEffect#removing-unnecessary-function-dependencies
https://react.dev/reference/react/useEffect#removing-unnecessary-object-dependencies

그리고 React에서는 eslint-plugin-react-hooks 플러그인을 통해 useEffect의 의존성 배열을 자동으로 검사하고, 누락된 의존성을 경고해주는 규칙인 react-hooks/exhaustive-deps를 제공하고 있는데, 이것도 활용하고 있는지 체크해보세요.

wonsss commented on components/src/lib/Modal.tsx:
```diff
@@ -0,0 +1,181 @@
+const ModalContext = createContext<(() => void) | undefined>(undefined);
+
+const Modal = ({ show, onHide, background = true, position = "center", gap = 16, children }: ModalProps) => {
+  useKeyEscClose(onHide);
```
현재 키보드 이벤트 처리를 통한 키보드 접근성 향상이 Escape 키만 있는데, 일반적으로 모달에 기대되는 사용성 등은 다음과 같습니다.

참고만 해주셔도 되고, 여력이 되시면 추가 구현해보실 수도 있을 것 같아요.

- 포커스 관리
모달이 열릴 때 자동으로 모달 내부로 포커스되어야 합니다.
모달이 닫힐 때 이전 포커스 위치로 돌아가야 합니다.

- Tab 키 관리
Tab 키로 모달 내부 요소들 간 이동할 수 있고 내부 요소들 이동으로 제한할 수 있어야 합니다.
Enter 키나 Space 키로 버튼 클릭이 처리되어야 합니다.

W3에서 제공하는 Modal Dialog 예시를 직접 실행해보시면, 위 설명을 이해하는 데 도움이 되실거예요.

https://www.w3.org/WAI/ARIA/apg/patterns/dialog-modal/examples/dialog/

jeongyou replied:
모달의 사용성 측면은 아직 깊이 고민하지 못했는데, 말씀해주셔서 감사합니다!

![image](https://github.com/user-attachments/assets/a1e83a27-766c-4690-8331-0f9138993d02)

보니까 step2 요구사항에 해당 내용이 포함되어 있더라고요!
알려주신 예시 자료도 참고해서, 포커스 이동 및 키보드 접근성 관련된 부분은 step2에서 반영해보도록 하겠습니다!

wonsss replied:
다음 요구사항이군요! 좋습니다!!

wonsss commented on components/src/lib/Modal.tsx:
```diff
@@ -0,0 +1,181 @@
+import { css } from "@emotion/react";
+import useKeyEscClose from "./useKeyEscClose";
+
+export interface ChildrenProps {
+  /** 자식 요소 (JSX.Element 또는 문자열 등) */
+  children: React.ReactNode;
+}
+
+export interface ModalProps extends ChildrenProps {
+  /** 모달을 보여줄지 여부 */
+  show: boolean;
+
+  /** 모달을 닫는 함수 (배경 클릭이나 X 버튼 클릭 시 호출) */
+  onHide: () => void;
+
+  /** 배경 어두움 여부  */
+  background?: boolean;
+
+  /** 모달 위치 설정: 가운데(center), 상단(top), 하단(bottom) */
+  position?: "center" | "top" | "bottom";
+
+  /** 모달 내부의 flex 간격 설정 */
+  gap?: number;
+}
+
+export interface ModalHeaderProps extends ChildrenProps {
+  /** 닫기 버튼(X)을 표시할지 여부 */
+  closeButton?: boolean;
+}
+
+export interface ModalTitleProps extends ChildrenProps {
+  /** 텍스트 색상 설정 */
+  color?: string;
+}
```
params에 JSDoc 주석을 잘 달아주셨네요! 👍

wonsss commented on components/src/stories/Modal.stories.tsx:
```diff
@@ -0,0 +1,57 @@
+
+type Story = StoryObj<typeof Modal>;
+
+export const ModalComponent: Story = {
+  render: (args, { updateArgs }) => {
```
스토리북의 상호작용 테스트를 활용해보시면 좋을 것 같아요. 실제 사용자의 행동을 시뮬레이션하여 컴포넌트가 예상대로 동작하는지 확인하는 데 매우 유용합니다.
테스트코드 play 함수를 스토리 객체에 추가하면, 스토리북에서 해당 컴포넌트의 상호작용을 자동으로 테스트할 수 있어요.

참고: https://storybook.js.org/docs/writing-tests/component-testing

```ts
export const MyStory: Story = {
  // 1. 렌더링 함수
  render: () => <MyComponent />,
  
  // 2. 상호작용 테스트
  play: async ({ canvasElement, step }) => {
    const canvas = within(canvasElement);
    
    await step('테스트 단계 설명', async () => {
      // 테스트 코드
    });
  }
};
```

wonsss commented on components/src/lib/Modal.tsx:
```diff
@@ -0,0 +1,181 @@
+  color?: string;
+}
+
+const ModalContext = createContext<(() => void) | undefined>(undefined);
```
undefined를 초기값으로 사용하고 있어서, Context를 사용하는 컴포넌트에서 불필요하게 undefined 체크가 필요한 상황이에요.
즉, 타입이 (() => void) | undefined로 되어있어, 실제로는 함수만 필요한데도 undefined를 처리해야 해요. 또한 현재 Context의 값이 단순 함수 하나만 전달되고 있어서, 나중에 다른 값이 필요할 때 확장이 어려워요.

이러한 문제를 해결하기 위해, 예를 들어, 다음과 같이 명시적으로 수정하실 수 있어요.

```ts
interface ModalContextType {
  onHide: () => void;
  // 나중에 필요한 다른 값들도 쉽게 추가
}

const ModalContext = createContext<ModalContextType>({
  onHide: () => {
    throw new Error('ModalContext must be used within a ModalProvider');
  }
});
```

jeongyou replied:
기존에는 Context의 초기값으로 `undefined`를 사용하고 있어서 불필요한 null 체크가 필요하고 Context 값이 단순히 함수 하나만 전달되는 구조라 나중에 확장이 어렵다는 문제가 있었네요

이 부분을 개선하기 위해 알려주신대로 명시적인 타입 객체(`ModalContextType`)를 선언하고, Context의 초기값을 `{ onHide: () => { ... } }` 형태로 변경하여 타입 안정성과 확장 가능성을 확보했습니다!

관련 커밋:

❓그리고 한가지 궁금한 점이 있습니다!

```ts
const ModalContext = createContext<ModalContextType>({
  onHide: () => {
    throw new Error('ModalContext must be used within a ModalProvider');
  }
```
이 부분에서 throw new Error(...)를 사용하신 이유가 궁금합니다...!
조금 찾아보면서 ModalContext.Provider 없이 useContext에 사용할 경우, 실수로 onHide를 호출하게 될 때 기본 함수가 실행되어 명시적인 에러를 발생시키기 위한 안전장치라고 이해했는데요, 제가 이해한 것이 맞는지 궁금해서 여쭤봅니다

wonsss replied:
네. 이해한 내용이 맞습니다!

wonsss commented on components/src/lib/Modal.tsx:
```diff
@@ -0,0 +1,181 @@
+  return (
+    <div css={ModalHeaderStyle}>
+      <span>{children}</span>
+      {closeButton && (
+        <div css={ModalCloseStyle} onClick={onHide}>
```
버튼이면, div보다 button 요소를 사용하는 것이 좋아요. 혹시 div를 사용하신 이유가 있으실까요?
태그 선택 시, 적절한 시맨틱 태그를 사용하는 것이 왜 중요할까요?

jeongyou replied:
사실 특별한 이유가 있었던 건 아니고... 평소에 별 생각 없이 div에 습관처럼 작성했던 것 같아요 😅
리뷰를 보코 시멘틱 태그를 사용하는게 중요한 이유를 찾아보았는데,
- button 요소를 쓰면 기본적으로 Tab 키 이동 및 Enter/Space 키 동작을 지원해줘서 따로 설정하지 않아도 되고,
- 태그만 봐도 "이건 클릭 가능한 요소구나"라고 직관적으로 이해할 수 있어서 협업과 유지보수에도 더 좋겠다는 생각이 들었습니다!

앞으로는 역할에 맞는 태그를 의식적으로 더 잘 선택해야겠네요...! 이 부분 반영하여 코드 수정하였습니다.

관련 커밋:

wonsss commented on components/src/App.tsx:
```diff
@@ -1,10 +1,15 @@
   return (
     <>
-      <h1>Component Modules</h1>
+      <Modal>
+        <Modal.Header closeButton>
+          <Modal.Title>Title</Modal.Title>
+        </Modal.Header>
+        <Modal.Body>Body</Modal.Body>
+        <Modal.Footer>Footer</Modal.Footer>
+      </Modal>
```
> 모달 컴포넌트 자유도
모달 컴포넌트를 제작할 때, 사용자에게 어느 정도의 자유도를 제공해야 할지에 대해 고민이 많았습니다. 처음에는 레이아웃과 스타일을 고정하고, 버튼이나 색상 등도 내부에서 모두 제어하려 했습니다. 하지만 그렇게 되면 사용자 입장에서 확장성이 떨어지고, 오히려 직접 모달을 만드는 편이 더 낫다고 느낄 수도 있겠다는 생각이 들었습니다.
반대로 너무 많은 자유도를 주면 props의 수가 급격히 늘어나고, 코드가 복잡해질 것 같았습니다. 그래서 저희는 필수 props만 받되, 나머지 영역은 children과 합성 컴포넌트 방식으로 구성하여 사용자에게 구조적 유연성을 제공하는 방식으로 설계했습니다.
모달 컴포넌트를 설계할 때, 어느 정도의 자유도를 허용하는 것이 바람직한지 궁금합니다.

너무 잘 해주셨어요 👍  컴파운드 컴포넌트 패턴을 적절히 잘 사용해주셨네요.
필수 로직은 props로 제어 가능하게 하여 기본 구조는 안내하되, 구성은 children과 Compound Component로 위임하여 사용자가 자유롭게 수정하고 확장 가능해졌네요.

스타일 커스터마이징은  props로 color, width 등 세세하게 다 넣는 것은 지양하고 className, css props를 받도록  제공하는 게 좋아요.

jeongyou replied:
오, 이 부분도 저희가 의도했던 방향입니다!
모달은 정말 기본 구조만 제공하고, Header, Body, Footer 같은 내부 구성은 사용자가 자유롭게 children과 합성 컴포넌트를 통해 조립할 수 있도록 설계했습니다!

그리고 스타일 커스터마이징에 대해 주신 피드백도 정말 몰랐던 부분이에요. color, width 같은 건 전부 props로 하나하나 받아야 하나…? 하고 고민하고 있었는데, className, css props를 받는 방식이 있었군요! 

찾아보니까
```tsx
const Modal = ({ children, className, css }) => (
  <div className={className} css={css}>
    {children}
  </div>
);
```

이런 방식으로 사용하는 거네요! 마르코 덕분에 많이 배워갑니다 🙏

wonsss commented on components/src/lib/Modal.tsx:
```diff
@@ -0,0 +1,181 @@
+
+const Modal = ({ show, onHide, background = true, position = "center", gap = 16, children }: ModalProps) => {
+  useKeyEscClose(onHide);
+  return (
+    <ModalContext.Provider value={onHide}>
+      <div css={ModalWrapperStyle(show)}>
```
React Portal을 사용하여 Modal을 구현해보는 방법을 한번 알아보시면 좋을 것 같아요.
왜 Modal 구현 시 Portal을 사용할까요? 

https://ko.react.dev/reference/react-dom/createPortal

jeongyou replied:
이번에 Portal을 처음 알게 되어서 모달에 자주 사용되는 이유를 정리해보았습니다.

일반적으로 모달을 DOM 트리 내에 렌더링하면, 부모 요소의 position, z-index, overflow 스타일에 영향을 받아 의도한 위치나 계층에서 제대로 렌더링되지 않는 문제가 생길 수 있다고 합니다. 하지만 포털을 사용하면 이러한 제약을 피하고, <body>처럼 전역적인 위치에 직접 렌더링할 수 있어서 레이아웃과 관계없이 안정적으로 표시할 수 있다는 장점이 있다고 하네요!

또한 접근성 측면에서도 이점이 있다고 하더라고요. 예를 들어 모달이 열려 있을 때 키보드로 Tab을 눌렀을 때 포커스가 모달 외부로 빠지지 않게 제한하는 포커스 트랩(focus trap) 구현에 유리하다고 해서, 이 부분도 함께 공부해보려고 합니다.

Portal에 대해서는 조금 더 공부해보면서, Step2에서 적용하면 좋을 것 같습니다! 감사합니다!👍

wonsss replied:
잘 정리해주셨네요! 👍

wonsss reviewed (changes requested):
안녕하세요! 기린 😄  반갑습니다!
페이먼츠 모듈 미션 1단계 수고 많으셨어요.
모달에서 컴파운드 컴포넌트 패턴을 사용하신 점이 인상적이었어요. 👍 
커스텀 훅은 입력 값 상태 관리가 아닌 유효성 검사에 집중하여 구현해주셨는데, 코멘트 참고해주세요!
질문에는 관련 코드 맥락에 답변을 달았어요.
리뷰 반영 후 재리뷰 요청주세요. 감사합니다 🙇

jeongyou commented on .github/workflows/chromatic.yml:
```diff
@@ -0,0 +1,25 @@
+# Workflow name
+name: 'Chromatic Deployment'
+
+# Event for the workflow
+on: push
+
+# List of jobs
+jobs:
+  chromatic:
+    name: 'Run Chromatic'
+    runs-on: ubuntu-latest
+    # Job steps
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
+      - name: Install dependencies
+        run: |
+          cd components
+          yarn install
+      - uses: chromaui/action@latest
+        with:
+          projectToken: ${{ secrets.CHROMATIC_PROJECT_TOKEN }}
+          token: ${{ secrets.GITHUB_TOKEN }}
+          workingDir: ./components
```
감사합니다~ㅎㅎ

jeongyou commented on components/src/lib/Modal.tsx:
```diff
@@ -0,0 +1,181 @@
+      <span>{children}</span>
+      {closeButton && (
+        <div css={ModalCloseStyle} onClick={onHide}>
+          <svg width="15" height="14" viewBox="0 0 15 14" fill="none" xmlns="http://www.w3.org/2000/svg">
+            <path
+              d="M14.4922 1.41L13.0822 0L7.49219 5.59L1.90219 0L0.492188 1.41L6.08219 7L0.492188 12.59L1.90219 14L7.49219 8.41L13.0822 14L14.4922 12.59L8.90219 7L14.4922 1.41Z"
+              fill="black"
+            />
+          </svg>
```
이 방법도 현업에서 많이 사용된다니 다행입니다! 저도 하드코딩된 점이 조금 신경 쓰였는데, 별도 컴포넌트로 분리해서 사용하는 방식은 생각 못 했네요. 좋은 팁 감사합니다! 😊

jeongyou committed:
refactor: cardNumberValidation 훅을 입력값 상태(inputStates) 포함한 입력 단위 훅으로 구조 전환

- inputStates 상태를 훅 내부에 추가하여 입력값을 직접 관리하도록 개선
- validateInput 제거, onChange(index)(e) 커링 구조로 입력 처리 통합
- errors 상태는 내부 전용으로 캡슐화하여 외부에 노출하지 않음
- 테스트 코드도 onChange 기반으로 리팩토링

jeongyou committed:
refactor: 입력 훅 구조에 맞게 inputStates 및 onChange 타입 개선

jeongyou committed:
refactor: useCvcNumberValidation 훅을 단일 입력 상태 중심 구조로 전환

- validateInput 제거, onChange(index)(e) 커링 구조로 입력 처리 통일
- inputStates 상태를 내부에서 관리하도록 개선
- errors 상태는 내부 전용으로 캡슐화, 외부에는 noError로 상태 전달
- 이에 따라 테스트 코드 전면 리팩토링 및 에러 메시지 검증 테스트 추가

jeongyou committed:
refactor: useExpirationDateValidation 훅을 입력값 상태(inputStates) 포함한 입력 단위 훅으로 구조 전환

- inputStates 상태를 훅 내부에 추가하여 월/년 입력값을 직접 관리하도록 개선
- validateInput 제거, onChange(index)(e) 커링 구조로 입력 처리 통합
- errors 상태는 내부 전용으로 캡슐화하여 외부에 노출하지 않음
- 테스트 코드도 onChange 기반으로 리팩토링하고 정상 입력 케이스 테스트 추가

jeongyou committed:
refactor: usePasswordValidation 훅을 입력값 상태(inputStates) 포함한 입력 단위 훅으로 구조 전환

- inputStates 상태를 훅 내부에 추가하여 월/년 입력값을 직접 관리하도록 개선
- validateInput 제거, onChange(index)(e) 커링 구조로 입력 처리 통합
- errors 상태는 내부 전용으로 캡슐화하여 외부에 노출하지 않음
- 테스트 코드도 onChange 기반으로 리팩토링하고 정상 입력 케이스 테스트 추가

jeongyou commented on hooks/src/lib/Hooks/useCardNumberValidation/index.ts:
```diff
@@ -0,0 +1,42 @@
+  ValidInputFuncType,
+} from '../../types';
+
+const useCardNumberValidation = (): ValidationHookReturnType => {
```
헉 이 미션의 핵심 의도를 제대로 이해하지 못하고 있었네요…! 😅
생각해보니, validateInput을 외부에 제공해서 사용자가 직접 검증하도록 하는 방식보다는,
훅 내부에서 값 저장 + 유효성 검사 + 에러 처리까지 모두 수행하도록 구성하는 것이 맞겠다는 판단이 들었습니다.

그래서 4개의 훅 모두 "입력 필드의 상태 관리 + 유효성 검사 로직"을 함께 포함하도록 리팩터링했고,
각 훅은 value 상태를 내부에서 관리하고, 입력 값 변경 핸들러만 외부에 제공하는 형태로 수정해보았습니다!

관련 커밋:
- [refactor: cardNumberValidation 훅을 입력값 상태(inputStates) 포함한 입력 단위 훅으로 구…](https://github.com/woowacourse/react-modules/pull/111/commits/5e6243204fbbe8a18c907005a5826c769b177433) 
- [refactor: useCvcNumberValidation 훅을 단일 입력 상태 중심 구조로 전환](https://github.com/woowacourse/react-modules/pull/111/commits/2c71f6b5638b4a5c76a624001925ef9ea9e8858d) 
- [refactor: useExpirationDateValidation 훅을 입력값 상태(inputStates) 포함한 입력 단…](https://github.com/woowacourse/react-modules/pull/111/commits/5f0aaa716daefea0c06d0567e300209c6311b782) 
- [refactor: usePasswordValidation 훅을 입력값 상태(inputStates) 포함한 입력 단위 훅으로 …](https://github.com/woowacourse/react-modules/pull/111/commits/8f212b33e68905259276bd22a246c32fa8e1474c)

wonsss replied:
잘 수정해주셨네요! 👍

jeongyou committed:
refactor: 단일 입력 훅들을 파생값 기반 구조로 리팩터링

- inputStates만 상태로 관리하고, error와 errorMessage는 입력값 기반으로 계산하도록 변경
- errors 상태 및 setErrors 제거로 상태 수 줄임

jeongyou commented on hooks/src/lib/Hooks/useCardNumberValidation/index.ts:
```diff
@@ -0,0 +1,42 @@
+    false,
+    false,
+  ]);
+  const [errorMessage, setErrorMessage] = useState<ErrorMessageType>('');
```
말씀 주신 “에러 상태는 입력값에 따라 계산 가능한 파생값이니 굳이 상태로 두지 않아도 된다”는 부분은 단일 입력 필드의 경우에는 충분히 가능한 구조라고 생각되어 리팩터링 진행해보았습니다!
하지만  카드번호 검증 훅이나 유효기간 검증 훅은 여러 개의 입력 필드를 다루고 있어서, 각 필드에 개별 에러 스타일을 적용하거나, 어느 필드에서 에러가 발생했는지를 파악하려면 errors를 배열 형태의 상태로 유지하는 편이 더 명확하고 관리하기 쉽다고 생각합니다!

그래서 다중 입력 필드 훅에서는 기존처럼 errors[] 상태를 유지하고, 단일 필드 훅은 입력값 기반으로 에러를 계산하는 구조로 코드 수정하였습니다! 😊

관련 커밋: [refactor: 단일 입력 훅들을 파생값 기반 구조로 리팩터링](https://github.com/woowacourse/react-modules/pull/111/commits/94ea02c6348b3994a35878651602eb2125742348)

wonsss replied:
네! 수고하셨습니다 👍 
참고로 공식문서 관련 내용 링크 공유드려요.
https://react.dev/learn/choosing-the-state-structure#avoid-redundant-state

jeongyou committed:
refactor: isNumber 함수명을 validateNumericString으로 명확하게 변경

jeongyou commented on hooks/src/lib/utils/validation.ts:
```diff
@@ -0,0 +1,32 @@
+
+  if (error) return { error, message: ERROR_MESSAGE.NUMBER_ONLY };
+  return { error, message: '' };
+};
```
말씀해주신 부분 생각해보니 isNumber 함수는 boolean을 반환할 것처럼 보이네요...!

그래서 함수명을 validateNumericString으로 변경하여 해당 함수가 숫자 유효성 검사 로직을 포함한 함수임을 더 명확하게 표현해보았습니다!

관련 커밋: [refactor: isNumber 함수명을 validateNumericString으로 명확하게 변경](https://github.com/woowacourse/react-modules/pull/111/commits/c7dd80b2b9fbe069515bfc404db0bb4efa90b8d6)

함수의 역할과 반환 타입을 좀 더 직관적으로 전달할 수 있게 되어 좋네요~~ 감사합니다!

jeongyou committed:
refactor: 숫자 유효성 검사 로직을 긍정 조건 기반으로 개선하여 가독성 향상

- 부정 조건(!) 대신 긍정 조건 변수(isValid)를 사용해 로직 명확화
- error 변수명이 혼동을 줄 수 있어 isValid로 변경

jeongyou committed:
refactor: 유효기간 검증 로직을 month/year 단위로 함수 분리하여 명확성 개선

- isExpirationDate 내부에서 validateMonth, validateYear 함수로 검증 로직을 위임
- 타입별 조건 분기를 명시적으로 처리하여 가독성과 재사용성 향상

jeongyou commented on hooks/src/lib/utils/validation.ts:
```diff
@@ -0,0 +1,32 @@
+  }
+
+  return { error: false, message: '' };
+};
```
유효기간 검증 로직을 validateMonth, validateYear로 분리해보았습니다! 분리를 해보니 함수의 역할이 더 명확하게 구분되고, 나중에 각각의 검증 규칙이 달라질 경우에도 확장성이 높아질 것 같습니다. 그리고 isExpirationDate가 더 읽기 쉬워진 것 같아 좋습니다!

관련 커밋: [refactor: 유효기간 검증 로직을 month/year 단위로 함수 분리하여 명확성 개선](https://github.com/woowacourse/react-modules/pull/111/commits/7c3fdbd29cc17989015ac39d9683125ac1319843)

jeongyou committed:
refactor: validateMonth, validateYear 함수 분리 후 validateExpirationDate 객체로 통합 export

- validateMonth, validateYear를 개별 함수로 분리해 테스트 가능성과 재사용성 확보
- 동시에 validateExpirationDate 객체로 묶어 네임스페이스처럼 사용 가능하도록 구성

jeongyou committed:
refactor: 유효기간 입력 전용 타입 정의 및 ValidationHookReturnType 분리

- ExpirationDateInputs, ExpirationDateErrors, ExpirationDateValidationReturnType 타입 신설
- month/year 구조를 명확히 표현하고 유효기간 전용 훅에서 사용하기 위한 타입 세분화
- CurriedInputChangeHandler 대신 ExpirationDateChangeHandler 도입
- 기존 공통 타입과 명확히 분리하여 역할별 책임을 구체화함

jeongyou committed:
refactor: checkNoError 유틸 함수가 객체 기반 에러 구조도 처리할 수 있도록 확장

- 기존 boolean 및 배열(boolean[])에 더해 객체 형태(ExpirationDateErrors)도 검사 가능하도록 로직 분기 추가
- 입력값의 타입 다양성을 수용해 다양한 훅의 에러 처리 방식에 공통 대응 가능

jeongyou committed:
refactor: 유효기간 검증 훅을 month/year 기반의 객체 상태로 구조 전환

- inputStates와 errors를 각각 month/year 키를 갖는 객체 형태로 변경하여 명시적 관리
- index 기반 분기 대신, type 기반 유효성 검증 함수(validateExpirationDate[type])로 처리
- 에러 메시지 상태는 입력 변경 시 마지막 검증 결과 기준으로 갱신
- 테스트 코드도 객체 구조와 validateExpirationDate 유틸 함수에 맞춰 수정

jeongyou commented on hooks/src/lib/Hooks/useExpirationDateValidation/index.ts:
```diff
@@ -0,0 +1,38 @@
+  const [errorMessage, setErrorMessage] = useState<ErrorMessageType>('');
+
+  const validateInput: ValidInputFuncType = (value: string, index: number) => {
+    const inputType = index === 0 ? 'month' : 'year';
```
처음에는 다른 훅들도 전부 index 기반으로 관리하고 있어서 유효기간도 당연히 index로 처리했었는데요, 피드백을 받고 다시 구조를 살펴보니 유효기간은 month, year처럼 명확한 의미를 가진 필드이고, 구성도 고정되어 있어서 굳이 index에 의존할 필요 없이 key 기반으로 관리하는 게 훨씬 더 명확하고 안전하겠다고 생각하게 되었습니다!

사용자 입장에서도 "0번은 월, 1번은 년"처럼 암묵적인 규칙을 기억하지 않아도 되고, 명시적인 key (month, year)를 직접 지정해 사용할 수 있어 사용성과 유지보수 측면에서도 훨씬 직관적인 구조가 되었다고 생각합니다!

그리고 검증 함수도 validateExpirationDate.month, validateExpirationDate.year처럼 유틸 객체로 명시적으로 분리해서 호출하도록 변경해서 훅 내부에서도 로직 흐름을 한눈에 들어오게, 더 명확하게 변경해보았습니다!

혹시 이 부분에서 제가 잘못 이해했거나 더 나은 방법이 있다면 피드백 부탁드립니다 🙏

관련 커밋: [refactor: 유효기간 검증 훅을 month/year 기반의 객체 상태로 구조 전환](https://github.com/woowacourse/react-modules/pull/111/commits/bdf436ff4c86fda437c1440f6fea70874c293920)

이 커밋 관련한 다른 커밋도 첨부합니다~
- [refactor: validateMonth, validateYear 함수 분리 후 validateExpirationDate …](https://github.com/woowacourse/react-modules/pull/111/commits/6519fed05c6b8c26432e24ce28330f6a30559421)
- [refactor: 유효기간 입력 전용 타입 정의 및 ValidationHookReturnType 분리](https://github.com/woowacourse/react-modules/pull/111/commits/1bed0e6a41ae41f3f11269729f69ccbea1901ba4)
- [refactor: checkNoError 유틸 함수가 객체 기반 에러 구조도 처리할 수 있도록 확장](https://github.com/woowacourse/react-modules/pull/111/commits/242ec1ff07425ba9c6e65172ec9c76b3c2fb86a2)

wonsss replied:
정확하게 이해해주셨고, 잘 정리해주시고 반영해주셨네요! 👍

jeongyou committed:
refactor: 카드 번호 입력값에 자릿수 유효성 검사 추가 및 테스트 보완

- validateCardNumber 함수에서 숫자 검증 후 자릿수 검증을 추가하여,
- 올바른 길이의 숫자가 입력되지 않을 경우 에러 메시지를 반환하도록 개선

jeongyou committed:
refactor: CVC 입력값에 대한 길이 및 숫자 검증 로직 개선

- 빈 문자열일 경우 초기 에러 상태가 되지 않도록 예외 처리 추가
- 숫자 외 입력 또는 길이 3이 아닌 경우 에러 메시지 반환

jeongyou committed:
feat: 유효기간 입력값에 대해 자리수(2자리) 검증 로직 추가

jeongyou committed:
feat: 카드 비밀번호 입력값에 길이(2자리) 검증 로직 추가

jeongyou commented on hooks/src/lib/Hooks/useCvcNumberValidation/index.ts:
```diff
@@ -0,0 +1,33 @@
+  const [errorMessage, setErrorMessage] = useState<ErrorMessageType>('');
+
+  const validateInput: ValidInputFuncType = (value: string) => {
+    const { error, message } = isNumber(value);
```
적용 후 위에 코멘트에 같이 달아 놓았습니다!

[refactor: CVC 입력값에 대한 길이 및 숫자 검증 로직 개선](https://github.com/woowacourse/react-modules/pull/111/commits/1268ea292c2499066d6be9cb9db0b73b0a03461d)

jeongyou commented on hooks/src/lib/Hooks/useCardNumberValidation/index.ts:
```diff
@@ -0,0 +1,42 @@
+  const [errorMessage, setErrorMessage] = useState<ErrorMessageType>('');
+
+  const validateInput: ValidInputFuncType = (value: string, index: number) => {
+    const { error, message } = isNumber(value);
```
적용 후 위에 코멘트에 같이 달아 놓았습니다!

[refactor: 카드 번호 입력값에 자릿수 유효성 검사 추가 및 테스트 보완](https://github.com/woowacourse/react-modules/pull/111/commits/b1a85211f69f54720d970c8a75be159c6629358e)

jeongyou committed:
refactor: 스타일 관련 코드 분리

jeongyou committed:
refactor: 의존성 배열에 onHide 함수 추가

jeongyou committed:
refactor: ModalContext에 명시적 타입 추가 및 타입 안정성 개선

jeongyou committed:
refactor: 버튼 태그 div에서 button으로 변경

- 버튼 기본 스타일도 제거

jeongyou commented:
안녕하세요 마르코! 리뷰 요청이 늦어져 죄송합니다.🙇‍♀️

마르코의 꼼꼼한 피드백 덕분에 제가 놓치고 있었던 부분이 많았다는 것을 알게 되었고, 기존 코드에서 고려하지 못했던 개념들도 다시 점검해볼 수 있었습니다.
또한 더 나은 방향에 대한 제안을 통해 고민의 폭도 넓어졌고, 전반적으로 많은 도움이 되었습니다. 정말 감사드립니다! 🙌

**각 리뷰 코멘트에 대한 수정 사항은 반영하였고, 해당 댓글에 답변 남겨두었습니다.**

확인 한번 부탁드립니다. 
좋은 하루 보내세요!😊

jeongyou commented:
@coderabbitai 이 PR에서 반복되는 안 좋은 습관이나 안티패턴이 있다면 알려주세요. 페인포인트가 있다면 어떤 부분인지도 분석해 주세요.

wonsss commented on hooks/src/lib/utils/validation.ts:
```diff
@@ -0,0 +1,32 @@
+  const error = !NUMBER_REGEX.test(value);
+
+  if (error) return { error, message: ERROR_MESSAGE.NUMBER_ONLY };
+  return { error, message: '' };
```
잘 수정해주셨네요! 
저도 처음 공부할 때, 클린코드가 무엇일지 고민하며 정리해둔 투박한 내용이 있는데, 혹시나 조금이라도 도움되실까봐 공유드려요.
https://wonsss.github.io/javascript/clean-code-javascript/#10-%EB%B6%80%EC%A0%95-%EC%A1%B0%EA%B1%B4%EB%AC%B8-%EC%A7%80%EC%96%91%ED%95%98%EA%B8%B0

wonsss commented on hooks/src/lib/Hooks/usePasswordValidation/index.ts:
```diff
@@ -0,0 +1,32 @@
+  const [errorMessage, setErrorMessage] = useState<ErrorMessageType>('');
+
+  const validateInput: ValidInputFuncType = (value: string) => {
+    const { error, message } = isNumber(value);
```
> ❓1. 현재 길이 검증 로직은 입력이 자리수를 충족하지 않으면 곧바로 에러 메시지를 띄우고, 입력이 모두 완료된 이후에 사라지는 구조로 작성해보았는데요, 이렇게 되면 사용자가 정상적으로 입력 중일 때에도 에러가 계속 보여서, 오히려 과한 피드백처럼 느껴질 수도 있을 것 같다는 고민이 들었습니다.
물론 이런 방식도 나쁘다고는 생각하지 않고, 즉각적인 피드백이 필요한 경우엔 유효하다고 느껴지긴 했는데요! 이 부분을 고민하다 보니, 예를 들어 onBlur 시점에만 길이 검증을 수행하는 방식도 있다는 걸 알게 되었습니다.
혹시 이런 검증 타이밍에 대해 마르코는 어떤 접근이 더 적절하다고 생각하시는지 궁금합니다..!

정말 좋은 UX 고민이네요. 말씀하신 두 가지 방식 모두 상황에 따라 사용됩니다.
무엇이 더 낫다고 제가 답변드리는 것은 크게 의미가 없을 것 같고, 의도에 따라 적절한 방식을 택하게 되는 것 같아요.
나중에 현업에서도, 이처럼 UX와 개발 방향 선택지들을 고민하시고, 함께 협업하는 동료(기획자, 디자이너 등)에게 이러한 선택지들과 장단점을 먼저 제안해주시면, 협업하기 좋은 동료가 되실 것 같습니다!
그리고 더 나은 UX가 무엇일지는 다른 유명 서비스들의 페이먼츠 UX를 참고하시면 도움이 되실 것 같아요.

> ❓2. 그리고 현재 훅마다 길이 검증 로직이 들어 있는데요,카드 번호의 경우엔 format을 props로 받아서 각 input마다 유연하게 길이를 조정할 수 있게 했지만, CVC는 고정된 3자리라 내부에서 직접 지정해도 괜찮다고 판단했습니다.
하지만 비밀번호나 유효기간처럼 길이 규칙이 고정이라고 단정하기 어려운 값들에 대해서는 내부에서 길이를 지정해주는 지금 방식이 확장성 측면에서 조금 마음에 걸리는 것 같습니다.
그래서 이런 값들도 외부에서 길이를 전달받도록 해야 하나 고민이 되었는데, 혹시 마르코는 이런 경우에 어떻게 처리하는 게 더 좋다고 생각하시나요? 사용자에게 일관되게 format을 받는 방식이 나을지, 아니면 상황에 따라 내부에 고정값을 두는 방식도 괜찮을지 의견이 궁금합니다!

"비밀번호나 유효기간처럼 길이 규칙이 고정이라고 단정하기 어려운 값"이라고 판단하셨다면, 외부에서 길이를 전달받도록 수정하는 것도 적절할 것 같아요.  😄

wonsss reviewed (approved):
안녕하세요 기린 🦒   😄 
리뷰 반영 너무 잘 해주셨네요! 👍 
1단계는 여기서 마무리해도 좋을 것 같습니다.
2단계에서 뵈어요! 화이팅입니다 💪

wmakerjun commented:
@coderabbitai
You are a senior React engineer.

# Task
From this pull request’s code, identify **exactly one recurring code smell or anti-pattern** whose removal will most improve long-term maintainability and correctness.

# How to respond
- **Pattern name**  
- **Where it appears** (files + line ranges or clear examples)  
- **Why it hurts** (short reason + link to React/JS best-practice doc if relevant)  
- **How to fix** (concise code snippet or pseudo-diff)  
- **Why this is the highest-leverage fix** (≤ 2 sentences)

# Style
Respond in bullet points, no extra tips or unrelated suggestions.

# Language
Please write your response **in Korean**.

wmakerjun commented:
@coderabbitai  
You are a senior React engineer.

# Task  
From this pull request’s code, identify **exactly one place** where the code is written in an imperative style, and explain how to make it **more declarative**.

# How to respond  
Start your response with the following markdown title:  
`# 1. ✨ 선언형 UI로서 개선할 피드백 포인트`  

Then respond with:  
- **Imperative pattern name or symptom**  
- **Where it appears** (files + line ranges or clear examples)  
- **Why it’s not declarative** (+ link to React’s declarative UI principle if helpful)  
- **How to rewrite it declaratively** (code snippet or pseudo-diff)  
- **Why this change improves React code quality** (≤ 2 sentences)

# Style  
Respond in bullet points, no extra tips or unrelated suggestions.

# Language  
Please write your response **in Korean**.

wmakerjun commented:
@coderabbitai  
You are a senior React engineer.

# Task  
From this pull request’s code, identify **exactly one state value** that can be **derived from other state or props**, and explain how removing it would simplify the code.

# How to respond  
Start your response with the following markdown title:  
`# 2. 📉 상태 최소화를 위한 개선 포인트`

Then respond with:  
- **Derived state name**  
- **Where it appears** (files + line ranges or variable name)  
- **Why it’s redundant**  
- **AS-IS**: _Show the redundant state and its logic_  
- **TO-BE**: _Show how to derive it from existing values_  
- **Why this is the highest-leverage fix** (≤ 2 sentences)

# Style  
Respond in bullet points, and focus on one state-related improvement.

# Language  
Please write your response **in Korean**.

wmakerjun commented:
@coderabbitai  
You are a senior React engineer.

# Task  
From this pull request’s code, identify **exactly one component** whose responsibility is unclear or mixed, and explain how to refactor it to **separate concerns more clearly**.

# How to respond  
Start your response with the following markdown title:  
`# 3. 🧩 컴포넌트 책임 분리를 위한 리팩토링 포인트`

Then respond with:  
- **Component name or file**  
- **Where the responsibility is unclear**  
- **Why this hurts reusability and maintainability**  
- **AS-IS**: _Show the current (overloaded) component logic_  
- **TO-BE**: _Show a refactored structure or extraction example_  
- **Why this is the most impactful structural change** (≤ 2 sentences)

# Style  
Respond in bullet points, and provide clear before/after contrast.

# Language  
Please write your response **in Korean**.

wmakerjun commented:
@coderabbitai  
You are a senior React engineer.

# Task  
From this pull request’s code, identify **exactly one recurring mistake, anti-pattern, or code smell** that appears in multiple places and seems to reflect the author's coding habit.  
Explain what it is, why it may keep happening, and how the author can recognize and change this pattern.

# How to respond  
Start your response with the following markdown title:  
`# 4. 🔁 반복적으로 나타나는 안티패턴 또는 코드 습관 피드백`

Then respond with:
- **Pattern or smell name**  
- **Where it appears repeatedly** (files + line ranges or variable/component patterns)  
- **Why it likely keeps happening** (assumed thought process or misconception)  
- **AS-IS**: _Example of current usage or structure_  
- **TO-BE**: _Refactored or improved structure with reasoning_  
- **How the author can self-check next time** (one reflection or checklist question)

# Style  
Respond in bullet points, focused on identifying **a single recurring behavioral pattern**.

# Language  
Please write your response **in Korean**.

jeongyou commented:
@coderabbitai  
You are a senior React engineer.

# Task  
From this pull request’s code, identify **exactly one place** where the code is written in an imperative style, and explain how to make it **more declarative**.

# How to respond  
Start your response with the following markdown title:  
`# 1. ✨ 선언형 UI로서 개선할 피드백 포인트`

Then respond with:  
- **Imperative pattern name or symptom**
- **Where it appears** (files + line ranges or clear examples)  
- **Why it’s not declarative**  
- **AS-IS**: _Show the current (imperative) code snippet_  
- **TO-BE**: _Show the rewritten (declarative) version_  
- **Why this change improves React code quality** (≤ 2 sentences)

# Style  
Respond in bullet points, and include only one clear improvement point.

# Language  
Please write your response **in Korean**.

jeongyou commented:
@coderabbitai

jeongyou commented:
@coderabbitai

선언형 명령형이 뭔데

jeongyou commented:
@coderabbitai  
You are a senior React engineer.

# Task  
From this pull request’s code, identify **exactly one state value** that can be **derived from other state or props**, and explain how removing it would simplify the code.

# How to respond  
Start your response with the following markdown title:  
`# 2. 📉 상태 최소화를 위한 개선 포인트`

Then respond with:  
- **Derived state name**  
- **Where it appears** (files + line ranges or variable name)  
- **Why it’s redundant**  
- **AS-IS**: _Show the redundant state and its logic_  
- **TO-BE**: _Show how to derive it from existing values_  
- **Why this is the highest-leverage fix** (≤ 2 sentences)

# Style  
Respond in bullet points, and focus on one state-related improvement.

# Language  
Please write your response **in Korean**.

jeongyou commented:
@coderabbitai  
You are a senior React engineer.

# Task  
From this pull request’s code, identify **exactly one component** whose responsibility is unclear or mixed, and explain how to refactor it to **separate concerns more clearly**.

# How to respond  
Start your response with the following markdown title:  
`# 3. 🧩 컴포넌트 책임 분리를 위한 리팩토링 포인트`

Then respond with:  
- **Component name or file**  
- **Where the responsibility is unclear**  
- **Why this hurts reusability and maintainability**  
- **AS-IS**: _Show the current (overloaded) component logic_  
- **TO-BE**: _Show a refactored structure or extraction example_  
- **Why this is the most impactful structural change** (≤ 2 sentences)

# Style  
Respond in bullet points, and provide clear before/after contrast.

# Language  
Please write your response **in Korean**.

jeongyou commented:
@coderabbitai
const modalContext = useMemo(() => ({ onHide }), [onHide]);

이게 뭔말이야? 어떤 의미야?

jeongyou commented:
@coderabbitai

const Modal = ({ show, onHide, background = true, position = "center", gap = 0, children }: ModalProps) => {
  // 컨텍스트 생성 및 닫기 함수 제공
  const modalContext = useMemo(() => ({ onHide }), [onHide]);
  
  // ESC 키 이벤트 처리를 컴포넌트 내부에서 직접 관리
  useKeyEscClose(onHide);
  
  return (
    <ModalContext.Provider value={modalContext}>
      <div css={ModalWrapperStyle(show)}>
        <div css={backGroundStyle(background)} onClick={onHide}></div>
        <div css={ModalContainerStyle(position, gap)}>{children}</div>
      </div>
    </ModalContext.Provider>
  );
};


이 컴포넌트가 리렌더링되는 조건은 show가 바뀌었을 때 리렌더링이 되는데 onHide는 굳이 useMemo로 처리 안해도 되는거 아니야?

혹시 모달을 쓰고있는 상위 컴포넌트에서 모달이 아닌 다른 자식 컴포넌트에 상태가 바뀌어서 리렌더링 될 때 onHide의 재할당을 막는건가...?

jeongyou commented:
@coderabbitai  
You are a senior React engineer.

# Task  
From this pull request's code, evaluate how well it avoids unnecessary re-renders due to component structure or prop changes.  
If the code already minimizes unnecessary renders effectively, affirm that with a short explanation.  
If there's room for improvement, suggest one specific way to reduce unnecessary re-renders—only if the change provides clear performance value without overengineering.

# How to respond  
Start your response with the following markdown title:  
# ✨ 한글 제목: 불필요한 렌더링 방지 여부 평가

Then respond with:
- ✅ **현재 코드가 원칙에 잘 맞는 이유** (if applicable)  
- ⚠️ **개선 여지가 있다면, 가능한 방향** (optional)  
- **AS-IS**: _Show the current code_
- **TO-BE**: _Show a refactored example_
- **스스로 점검할 수 있는 질문 하나**

# Style  
Use bullet points only. Do not force a suggestion—only provide one if it adds meaningful value.

# Language  
Please write your response **in Korean**.
